### DISCUSSION TITLE
A framework based on Termux & Termux:X11, to port GUI apps from Linux to Android

### DISCUSSION THREAD

**Discussion by mister-teddy** [termux/termux-app#4068] [Ideas]
Just like how Electron wrapped Node.js and Chromium into each application which was built on it, could we just wrap Termux + Termux:X11 into every application which was built on this new framework?

It would be nice to have useful applications such as Visual Studio Code and Chromium as standalone applications, with their own icon on the app drawer or the home screen, one-touch install, one-touch launch, automatically resizable (if applicable), without knowing how they work under the hood. It would be a fun framework to maintain, but I want to make sure it is possible before putting all my effort into it. As a Termux/Android/Linux developer, **how feasible is it** do you think? And as an application user, how helpful would you think this framework will be? Thank you 🙏

**Comment by TomJo2000** [comment]
VSCode and Chromium are available in the [TUR](https://github.com/termux-user-repository/tur?tab=readme-ov-file#termux-user-repository-tur).
- https://github.com/termux-user-repository/tur/tree/master/tur/chromium
- https://github.com/termux-user-repository/tur/tree/master/tur/code-oss

You can make launcher icons for them using [Termux:Widget](https://github.com/termux/termux-widget) as per usual.

**Comment by twaik** [comment]
Are you talking about something like https://play.google.com/store/apps/details?id=tech.ula.gimp&hl=en ? The app which has only Gimp (with all dependencies), some WM only for one application and its windows and XServer XSDL packed into one application?

**Reply by twaik** [reply]
Probably it is doable, but it will not be comfortable to use it as is. It will require some effort to make Termux:X11 integrate to Android's windowing system.

**Reply by twaik** [reply]
Probably we could do that to earn some funds for hosting needs. GUI apps will not require terminal emulator and background services so we could simply start Termux:X11, minimal WM or minimal DE and target application. This application will not need package manager or something, all libraries will be built in into APK and will have no permissions problems (like we had with ALL_FILES permission in some versions on modern Android versions). It will be much easier to maintain this. 
Also these small applications will be able to be published to google play and serve some kind of Termux advertisement even if it is paid. And of course we can have it available free of charge on F-droid.

**Reply by twaik** [reply]
@agnostic-apollo what do you think about that?

**Reply by mister-teddy** [reply]
Yes, this is exactly what I meant. Do you know how they ported that GIMP app? Did they use Termux and Termux:X11 under the hood?

It seems like the main issue with these kind of apps is that they are hard to use on small screens. But my intent was to use them on large Android tablets (which I expect to be more relevant in the future), and Dex. So yes, work must be done to make them effortlessly fit in the Android application view.

**Reply by twaik** [reply]
They used some distro in proot environment and XServer XSDL. Not termux and termux-x11.

**Reply by mister-teddy** [reply]
Thank you for the information @twaik! Good to know that it is somewhat feasible.

**Reply by agnostic-apollo** [reply]
Node.js and Chromium being bundled into a single application makes sense because that's all that's required to run sites and dependencies are downloaded at runtime via the web.

Visual Studio is a completely different thing as its a development environment, and not an end product like electron apps. It also supports many languages and they each would have their own packages and then additional dependencies, depending on what is intended to be run, so you each such variant would require separate apps. Execution of dependencies downloaded at runtime would obviously be an issue, unless using linker hack if deploying on playstore or packing every required executable as native library.

https://github.com/agnostic-apollo/Android-Docs/blob/master/site/pages/en/projects/docs/apps/processes/app-data-file-execute-restrictions.md

Mobile devices often don't have enough space either for such duplications like desktops allow with electron, i.e each app would have duplicates of each Termux package. Since Termux is a rolling release model, might be better to just build framework inside Termux itself.

X11 apps in Termux would obviously have performance and usability issues as well, compared to native apps.

What you ask should theoretically be possible, whether its feasible is for you to decide yourself. There are already some apps on playstore that ship minimal termux packages for development with IDE support.

**Reply by mister-teddy** [reply]
Insightful, thank you!

**Reply by agnostic-apollo** [reply]
Welcome
